File: Session_04_GreedyAlgorithms/multi_language_implementations/fractional_knapsack.py
import sys

class Item:
    def __init__(self, item_id, value, weight):
        self.id = item_id
        self.value = value
        self.weight = weight
        self.ratio = 0.0

def solve_fractional_knapsack(items, capacity):
    for item in items:
        item.ratio = item.value / item.weight
        
    # Sort items by value/weight ratio descending
    items.sort(key=lambda x: x.ratio, reverse=True)
    
    total_value = 0.0
    remaining_capacity = capacity
    
    for item in items:
        if remaining_capacity <= 0:
            break
        if item.weight <= remaining_capacity:
            remaining_capacity -= item.weight
            total_value += item.value
        else:
            fraction = remaining_capacity / item.weight
            total_value += item.value * fraction
            remaining_capacity = 0
            
    print(f"Max Knapsack value: {total_value:.2f}")

def main():
    items = [
        Item(1, 60.0, 10.0),
        Item(2, 100.0, 20.0),
        Item(3, 120.0, 30.0)
    ]
    solve_fractional_knapsack(items, 50.0)
    print()
    input("Press Enter to exit...")

if __name__ == "__main__":
    main()

File: Session_04_GreedyAlgorithms/multi_language_implementations/fractional_knapsack.js
const readline = require('readline');

function solveFractionalKnapsack(items, capacity) {
    const itemsWithRatio = items.map(item => ({
        ...item,
        ratio: item.value / item.weight
    }));

    const sortedItems = itemsWithRatio.sort((a, b) => b.ratio - a.ratio);
    let totalValue = 0.0;
    let remainingCapacity = capacity;

    for (let item of sortedItems) {
        if (remainingCapacity <= 0) break;
        if (item.weight <= remainingCapacity) {
            remainingCapacity -= item.weight;
            totalValue += item.value;
        } else {
            let fraction = remainingCapacity / item.weight;
            totalValue += item.value * fraction;
            remainingCapacity = 0;
        }
    }
    console.log(`Max Knapsack value: ${totalValue.toFixed(2)}`);
}

function main() {
    const items = [
        {id: 1, value: 60.0, weight: 10.0},
        {id: 2, value: 100.0, weight: 20.0},
        {id: 3, value: 120.0, weight: 30.0}
    ];
    solveFractionalKnapsack(items, 50.0);
    console.log();

    const rl = readline.createInterface({
        input: process.stdin,
        output: process.stdout
    });
    rl.question("Press Enter to exit...", () => {
        rl.close();
    });
}

main();
